### PR TITLE
Adds in requisition tablet for Crashed

### PR DESCRIPTION
**Pull request by ma44**
## About The Pull Request
Adds in a tablet and a beacon to the canterbary. When attacking the beacon with the tablet, it will link the beacon to the tablet and allows delivery of items to said beacon provided the surrounding area doesn't have a ceiling or is a glass ceiling.
Fulton extraction kits now give supply points based off of the actual supply exports that the elevator uses instead of it's own thing, this should give a more overall buff to on site extraction and can be subjected to balance later on. 

## Why It's Good For The Game
Makes crashed marines have an option of refillable ammo and other goodies when they kill some xenos, this should be a good enough buff to more even out the xeno vs marine forces in the gamemode. 

## Changelog
:cl: ma44
add: The crashed gamemode's shuttle now has a requisitions tablet and beacon in the cockpit, attack the beacon with the tablet to enjoy the luxuries of requisitions right at your doorstep!
/:cl:

### CODE REVIEW

**Comment by jroinc** [comment]
On an average crash round, I counted 26 dead ayys, not counting gibbed ones.

Everyone's gonna be a B-18.

File: code/modules/shuttle/supply.dm [hunk 1]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
 
 	updateUsrDialog()
+
+//A tablet that can order requisition goodies; when it's linked to a beacon it will deliver the goods via launch pods nearby the beacon
+
+/obj/item/req_tablet
+	name = "requisition tablet"
+	desc = "A handheld tablet that allows the average user to remote order supply packages. When this is linked to a beacon, it will deliver the goods VIA launch pods. \
+			The pods will drop in a 5x5 square around the beacon and will not injure or explode, it will teleport away shortly after dropping off the contents."
+	var/obj/item/radio/beacon/linked_beacon
```
Comment (psykzz) [inline code/modules/shuttle/supply.dm]: You should move the new variables underneath parent vars

File: code/modules/requisitions/_supply_export.dm [hunk 2]
```diff
@@ -1,22 +1,7 @@
...
-			. = 10
-		if(XENO_TIER_THREE)
-			. = 30
-		if(XENO_TIER_FOUR)
-			. = 50
+	. = 0
+	for(var/datum/supply_export in GLOB.exports_types)
+		var/datum/supply_export/E = supply_export
+		if(type == E.export_obj)
+			. = E.cost
```
Comment (Rohesie) [inline code/modules/requisitions/_supply_export.dm]: This was intentionally changed so we could balance stuff on Crash separately, as xeno bodies are much cheaper to come back. Do please revert this change. Let's start small and if xenos are way too cheap we can always increase their value.

File: code/modules/shuttle/supply.dm [hunk 3]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+
+/obj/item/req_tablet/examine(mob/user)
+	. = ..()
+	if(!linked_beacon)
+		to_chat(user, "There seems to be no beacon attached to this tablet, hit a beacon with a tablet to link to it.")
+	else
+		to_chat(user, "There seems to be a beacon attached to this tablet. Should be good to go for delivery.")
+	return .
+
+/obj/item/req_tablet/afterattack(atom/A, mob/user, proximity)
```
Comment (psykzz) [inline code/modules/shuttle/supply.dm]: Shouldn't this be attack_by otherwise wont you hit them first then link it?
Actually hit it and take damage i mean

File: code/modules/shuttle/supply.dm [hunk 4]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+//Begin the requisitions copypasta
+/obj/item/req_tablet/interact(mob/user)
+	. = ..()
+	if(.)
+		return
+	var/dat
+	if(temp)
+		dat = temp
+	else
+		if (SSshuttle.supply)
```
Comment (psykzz) [inline code/modules/shuttle/supply.dm]: do an early return here instead

File: code/modules/shuttle/supply.dm [hunk 5]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+	var/temp //TGUI things
+	var/last_viewed_group = "categories"
+
+/obj/item/req_tablet/examine(mob/user)
+	. = ..()
+	if(!linked_beacon)
+		to_chat(user, "There seems to be no beacon attached to this tablet, hit a beacon with a tablet to link to it.")
+	else
+		to_chat(user, "There seems to be a beacon attached to this tablet. Should be good to go for delivery.")
+	return .
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]:
```suggestion
```

File: code/modules/shuttle/supply.dm [hunk 6]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+	if(temp)
+		dat = temp
+	else
+		if (SSshuttle.supply)
+			if(!linked_beacon)
+				dat = "<BR><B>WARNING: NO CONNECTIVITY WITH SUPPLY SYSTEM DUE TO LACK OF LINKED BEACON!!</B>"
+			else
+				dat += "<BR><B>Automated Launch Pod Retrieval System: Tablet edition</B>"
+				dat += "<BR><HR>Supply points: [round(SSpoints.supply_points)]<BR>"
+				dat += "<BR><HR><A href='?src=\ref[src];order=categories'>Order supply packs (CANNOT BE CANCELED ONCE ORDERED)</A>"
```
Comment (psykzz) [inline code/modules/shuttle/supply.dm]:
```suggestion
				dat += "<BR><HR><A href='?src=[REF(src)];order=categories'>Order supply packs (CANNOT BE CANCELED ONCE ORDERED)</A>"
```

File: code/modules/shuttle/supply.dm [hunk 7]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+	popup.set_content(dat)
+	popup.open(FALSE)
+	onclose(user, "computer")
+
+/obj/item/req_tablet/Topic(href, href_list)
+	. = ..()
+	if(.)
+		return
+
+	if(href_list["order"])
```
Comment (psykzz) [inline code/modules/shuttle/supply.dm]: make sure to return within each `if(href_list["blah"])`

File: code/modules/shuttle/supply.dm [hunk 8]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+	. = ..()
+	if(!linked_beacon)
+		to_chat(user, "There seems to be no beacon attached to this tablet, hit a beacon with a tablet to link to it.")
+	else
+		to_chat(user, "There seems to be a beacon attached to this tablet. Should be good to go for delivery.")
+	return .
+
+/obj/item/req_tablet/afterattack(atom/A, mob/user, proximity)
+	if(istype(A, /obj/item/radio/beacon))
+		to_chat(user, "You link the tablet to the beacon you hit it with, overriding the previous beacon linked if there was one.")
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]:
```suggestion
		if(A == linked_beacon)
			return
		to_chat(user, "<span class='notice'>You link the tablet to the beacon[linked_beacon ? " overriding the previous one!" : "!" ]</span>")
```

File: code/modules/shuttle/supply.dm (hunk 4 above)
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]: `else if`

File: code/modules/shuttle/supply.dm [hunk 9]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+	. = ..()
+	if(.)
+		return
+	var/dat
+	if(temp)
+		dat = temp
+	else
+		if (SSshuttle.supply)
+			if(!linked_beacon)
+				dat = "<BR><B>WARNING: NO CONNECTIVITY WITH SUPPLY SYSTEM DUE TO LACK OF LINKED BEACON!!</B>"
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]:
```suggestion
				dat = "<BR><B>WARNING: NO CONNECTIVITY WITH SUPPLY SYSTEM. LIKELY CAUSE: LACK OF LINKED BEACON.</B>"
```

File: code/modules/shuttle/supply.dm [hunk 10]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+			for(var/supply_group_name in GLOB.all_supply_groups)
+				temp += "<A href='?src=\ref[src];order=[supply_group_name]'>[supply_group_name]</A><BR>"
+		else
+			last_viewed_group = href_list["order"]
+			temp = "<b>Supply points: [round(SSpoints.supply_points)]</b><BR>"
+			temp += "<A href='?src=\ref[src];order=categories'>Back to all categories</A><HR><BR><BR>"
+			temp += "<b>Request from: [last_viewed_group]</b><BR><BR>"
+			for(var/supply_type in SSshuttle.supply_packs)
+				var/datum/supply_packs/N = SSshuttle.supply_packs[supply_type]
+				if(N.hidden || N.contraband || N.group != last_viewed_group) continue								//Have to send the type instead of a reference to
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]:
```suggestion
				if(N.hidden || N.contraband || N.group != last_viewed_group)
					continue								//Have to send the type instead of a reference to
```

File: code/modules/shuttle/supply.dm [hunk 11]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+			playsound(get_turf(src), 'sound/machines/triple_beep.ogg', 25, TRUE)
+			return
+		var/list/turfs_to_land_on = list()
+		for(var/turf/open/floor/T in range(2, linked_beacon))//Deliver around the beacon linked
+			var/area/A = get_area(T)
+			if(A.ceiling != CEILING_GLASS && A.ceiling != CEILING_NONE)
+				return
+			turfs_to_land_on += T
+
+		for(var/datum/supply_order/SO in SSshuttle.shoppinglist)
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]: Can there be a different type in this list?

File: code/modules/shuttle/supply.dm [hunk 12]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+		var/datum/supply_packs/P = SSshuttle.supply_packs[href_list["doorder"]]
+		temp = "Invalid Request"
+		temp += "<BR><A href='?src=\ref[src];order=[last_viewed_group]'>Back</A>|<A href='?src=\ref[src];mainmenu=1'>Main Menu</A>"
+		if(SSpoints.supply_points >= round(P.cost))
+			SSpoints.supply_points -= round(P.cost)
+			var/datum/supply_order/O = new /datum/supply_order()
+			O.id = SSshuttle.ordernum
+			O.pack = P
+			SSshuttle.shoppinglist += O
+			P.cost = P.cost * SUPPLY_COST_MULTIPLIER
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]:
```suggestion
			P.cost *= SUPPLY_COST_MULTIPLIER
```

File: code/modules/shuttle/supply.dm [hunk 13]
```diff
@@ -596,3 +589,135 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+			temp = "Thank you for your order.<BR>"
+			temp += "<BR><A href='?src=\ref[src];order=[last_viewed_group]'>Back</A>|<A href='?src=\ref[src];mainmenu=1'>Main Menu</A>"
+		else
+			temp = "Not enough supply points.<BR>"
+			temp += "<BR><A href='?src=\ref[src];order=[last_viewed_group]'>Back</A>|<A href='?src=\ref[src];mainmenu=1'>Main Menu</A>"
+
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]: Should this generate an order or return?

File: code/modules/shuttle/supply.dm (hunk 3 above)
Comment (ma44) [reply code/modules/shuttle/supply.dm]: Haven't gotten the beacon to take damage from hitting it with a tablet

File: code/modules/shuttle/supply.dm (hunk 11 above)
Comment (ma44) [reply code/modules/shuttle/supply.dm]: Nope

File: code/modules/shuttle/supply.dm (hunk 13 above)
Comment (ma44) [reply code/modules/shuttle/supply.dm]: No order should be made

File: code/modules/shuttle/supply.dm (hunk 7 above)
Comment (ma44) [reply code/modules/shuttle/supply.dm]: If it returns then updateUsrDialog() wouldn't get called; as i don't want to put that proc in every single href list then return i didn't put in return for each href

File: code/modules/requisitions/_supply_export.dm (hunk 2 above)
Comment (ma44) [reply code/modules/requisitions/_supply_export.dm]: Putting it back as it was before

**Comment by ma44** [comment]
Pushed new sprite @TheBimmer231

File: code/modules/requisitions/_supply_export.dm [hunk 14]
```diff
@@ -19,4 +19,3 @@
 			. = 30
 		if(XENO_TIER_FOUR)
 			. = 50
-	SSpoints.supply_points += .
```
Comment (Rohesie) [inline code/modules/requisitions/_supply_export.dm]: Intended?

File: code/modules/shuttle/supply.dm [hunk 15]
```diff
@@ -596,3 +594,139 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+/obj/item/req_tablet/interact(mob/user)
+	. = ..()
+	if(.)
+		return
+	var/dat
+	if(temp)
+		dat = temp
+	if(!SSshuttle.supply)
+		return
+	else
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]: No need for the `else`, the `if` returns already.

File: code/modules/shuttle/supply.dm (hunk 11 above)
Comment (ma44) [reply code/modules/shuttle/supply.dm]: Nope

File: code/modules/requisitions/_supply_export.dm (hunk 14 above)
Comment (ma44) [reply code/modules/requisitions/_supply_export.dm]: Nope, fixed

File: code/modules/shuttle/supply.dm (hunk 4 above)
Comment (ma44) [reply code/modules/shuttle/supply.dm]: No longer needed

File: code/modules/shuttle/supply.dm (hunk 3 above)
Comment (Rohesie) [reply code/modules/shuttle/supply.dm]: Maybe `preattack` then?

File: code/modules/shuttle/supply.dm (hunk 15 above)
Comment (Rohesie) [reply code/modules/shuttle/supply.dm]: Remove the `else` instead of removing the early `return`.

File: code/modules/shuttle/supply.dm (hunk 11 above)
Comment (Rohesie) [reply code/modules/shuttle/supply.dm]: Then remove the type casting inside the `for` loop to avoid a needless check.

**Comment by Rohesie** [comment]
Only two change requests pending.

File: code/modules/shuttle/supply.dm [hunk 16]
```diff
@@ -596,3 +594,141 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+
+/obj/item/req_tablet
+	name = "requisition tablet"
+	desc = "A handheld tablet that allows the average user to remote order supply packages. When this is linked to a beacon, it will deliver the goods VIA launch pods. \
+			The pods will drop in a 5x5 square around the beacon and will not injure or explode, it will teleport away shortly after dropping off the contents."
+	icon = 'icons/obj/items/req_tablet.dmi'
+	icon_state = "req_tablet_off"
+	var/obj/item/radio/beacon/linked_beacon
+	var/podType = /obj/structure/closet/supplypod/centcompod //No explosion, fast landing and teleports itself out of the landing destination
+	var/temp //TGUI things
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]: Give this a more informative name.

File: code/modules/shuttle/supply.dm [hunk 17]
```diff
@@ -596,3 +594,141 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+	var/last_viewed_group = "categories"
+
+/obj/item/req_tablet/examine(mob/user)
+	. = ..()
+	if(!linked_beacon)
+		to_chat(user, "There seems to be no beacon attached to this tablet, hit a beacon with a tablet to link to it.")
```
Comment (Rohesie) [inline code/modules/shuttle/supply.dm]:
```suggestion
		to_chat(user, "There seems to be no beacon attached, use the tablet on a beacon to link to it.")
```

File: code/modules/shuttle/supply.dm [hunk 18]
```diff
@@ -596,3 +594,141 @@ GLOBAL_LIST_EMPTY(exports_types)
...
+		if(A == linked_beacon)
+			return
+		to_chat(user, "<span class='notice'>You link the tablet to the beacon[linked_beacon ? " overriding the previous one!" : "!" ]</span>")
+		playsound(get_turf(src), 'sound/machines/triple_beep.ogg', 25, TRUE)
+		icon_state = "req_tablet_on"
+		. = TRUE
+	return ..()
+
+//Begin the requisitions copypasta
+/obj/item/req_tablet/interact(mob/user)
```
Comment (spookydonut) [inline code/modules/shuttle/supply.dm]: duplicating the supply code is dumb tbh

File: code/modules/shuttle/supply.dm (hunk 16 above)
Comment (spookydonut) [reply code/modules/shuttle/supply.dm]: its also not TGUI

**Comment by Rohesie** [comment]
Needs fixing, UI doesn't work and needs more feedback on how to be used.

**Comment by David-Stormwell** [comment]
Could you add a error message when in a bad place to call down supplies? Like "Error, can not deliver supplies due to roof in the way."

**Comment by psykzz** [comment]
@spookydonut @Rohesie did you have anything else on this PR?
Its been TM'd for a bunch of rounds already without issue, im thinking of merging it.

**Comment by Rohesie** [comment]
There's a single change request from me left. It's pretty simple.
Has ma44 stopped working on this?

**Comment by psykzz** [comment]
Yeah they were busy and asked for some time with their PRs

**Comment by Rohesie** [comment]
This looks functional, it just has code quality issues.
Up to you if you want to merge it.
Alternatively someone else can take it over, or maintainers can directly commit to the PR.

**Comment by spookydonut** [comment]
Don't merge code that has quality issues with some expectation that it'll get fixed later. - ninjanomnom, maybe occasions

**Comment by psykzz** [comment]
I didn't plan to merge. I wanted to get to grips with what your comments are since this is super old and not sure where it still stands.

**Comment by spookydonut** [comment]
reqs ui needs to be made into tgui then this can just just use that same tgui.

**Comment by Rohesie** [comment]
Not many people know how to do that. Is that something you'd be willing to?

**Comment by spookydonut** [comment]
> 
> 
> Not many people know how to do that. Is that something you'd be willing to?

Let's see how my upcoming discord ban appeal goes first
